users_app/serializers.py

```diff
@@ -0,0 +1,55 @@
+from rest_framework import serializers
+from django.contrib.auth import authenticate
+from .models import CustomUser
+
+
+class LoginSerializer(serializers.Serializer):
+    username = serializers.CharField()
+    password = serializers.CharField(write_only=True)
+
+    def validate(self, data):
+        username = data.get('username')
+        password = data.get('password')
+
+        user = authenticate(username=username, password=password)
+
+        if not user or not user.is_active:
+            raise serializers.ValidationError("Invalid credentials")
+
+        data['user'] = user
+        return data
+
+
+class RegisterSerializer(serializers.ModelSerializer):
+    password1 = serializers.CharField(write_only=True)
+    password2 = serializers.CharField(write_only=True)
+
+    class Meta:
+        model = CustomUser
+        fields = ["username", "password1", "password2", "first_name", "last_name", "email"]
+
+    def validate_username(self, value):
+        if CustomUser.objects.filter(username=value).exists():
+            raise serializers.ValidationError("Username is already taken.")
```

Comment:
"If your are concerned about privacy, regardless if the account was created or not make sure to respond the same way, and probably [204](https://httpstatuses.com/204) or [202](https://httpstatuses.com/202) are the most appropriated status code in this case.
To not confuse the user on the frontend you can display a generic message saying something like "You will receive a confirmation email on the next minutes if you don't have an account, if you don't receive the email try forget password".
Depending on how far you want to take things, you might want to create the account on a background process rather than in the main/request thread, otherwise attackers could analyze the response time of your endpoint and infer if the account was created or not based on the response time, this since the process of actually creating the account might take more time than just checking if it exists and returning.

**Responding the same way in both scenarios is the only way to ensure an attacker can't figure out who is already registered in your system.**" https://stackoverflow.com/questions/26587082/http-status-code-for-username-already-exists-when-registering-new-account